misk/testing/src/main/kotlin/misk/cloud/gcp/storage/InMemoryStorageRpc.kt

```diff
@@ -0,0 +1,394 @@
+package misk.cloud.gcp.storage
+
+import com.google.api.services.storage.model.Bucket
+import com.google.api.services.storage.model.BucketAccessControl
+import com.google.api.services.storage.model.Notification
+import com.google.api.services.storage.model.ObjectAccessControl
+import com.google.api.services.storage.model.Policy
+import com.google.api.services.storage.model.ServiceAccount
+import com.google.api.services.storage.model.StorageObject
+import com.google.api.services.storage.model.TestIamPermissionsResponse
+import com.google.cloud.Tuple
+import com.google.cloud.storage.StorageException
+import com.google.cloud.storage.spi.v1.RpcBatch
+import com.google.cloud.storage.spi.v1.StorageRpc
+import com.google.common.io.ByteStreams.toByteArray
+import java.io.InputStream
+import java.math.BigInteger
+import java.nio.file.FileAlreadyExistsException
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+internal class InMemoryStorageRpc : StorageRpc {
+    private val content = mutableMapOf<String, ByteArray>()
+    private val metadata = mutableMapOf<String, StorageObject>()
+    private val pendingContent = mutableMapOf<String, ByteArray>()
+    private val lock = ReentrantReadWriteLock()
```

Comment:
Is this because we have one VM-wide datastore and tests could be running in parallel?

Reply:
Usually you'll create a new instance per test so it's not an issue of multiple tests running in parallel. But you might use this for development mode services, in which case there could be multiple threads accessing this singleton concurrently.

Reply:
Ah, got it. Makes sense.

Reply:
because in multi-threading cases there will be way more readers than writers